File: testcigar.js

import fs from 'fs';
import { makeCigarTree, expandCigarTree, countGapSizes } from './cigartree.js';
import { parseHistorianParams, subLogLike, transLogLike, sum } from './likelihood.js';

if (process.argv.length != 5) {
    console.error('Usage: ' + process.argv[1] + ' tree.nh align.fa model.json');
    process.exit(1);
}

const [ treeFilename, alignFilename, modelFilename ] = process.argv.slice(2);
const treeStr = fs.readFileSync(treeFilename).toString();
const alignStr = fs.readFileSync(alignFilename).toString();
const modelJson = JSON.parse (fs.readFileSync(modelFilename).toString());

const ct = makeCigarTree (treeStr, alignStr);
const { alignment, expandedCigar, nodeName, distanceToParent, parentIndex, leavesByColumn, internalsByColumn, branchesByColumn } = expandCigarTree (ct);
const lcAlignment = alignment.map ((s) => s.toLowerCase());

const { alphabet, mixture, indelParams } = parseHistorianParams (modelJson);
const { transCounts } = countGapSizes (expandedCigar);

const { subRate, rootProb } = mixture[0];
const subll = subLogLike (lcAlignment, distanceToParent, leavesByColumn, internalsByColumn, branchesByColumn, alphabet, rootProb, { subRate });
const subll_total = sum (subll);

const transll = transLogLike (transCounts, distanceToParent, { indelParams, alphabet });
const transll_total = sum (transll);

console.log (JSON.stringify({'loglike':{'subs':subll_total,'indels':transll_total}, 'cigartree': ct}));

File: python/testcigar.py
import json
from jsonargparse import CLI

import jax.numpy as jnp

import cigartree
import likelihood
import h20

def main (modelFilename: str,
          treeFilename: str,
          alignFilename: str,
          ):
    """
    Compute the log likelihood of a tree given an alignment and a model.
    
    Args:
        treeFilename: Newick tree file
        alignFilename: FASTA alignment file
        modelFilename: Historian-format JSON file with model parameters
    """
    with open(modelFilename, 'r') as f:
        modelJson = json.load (f)
    with open(treeFilename, 'r') as f:
        treeStr = f.read()
    with open(alignFilename, 'r') as f:
        alignStr = f.read()

    ct = cigartree.makeCigarTree (treeStr, alignStr)

    alphabet, mixture, indelParams, *_others = likelihood.parseHistorianParams (modelJson)
    seqs, _nodeName, distanceToParent, parentIndex, transCounts = cigartree.getHMMSummaries (treeStr, alignStr, alphabet)

    assert len(mixture) == 1, "Only one mixture component is supported for substitution model"
    assert len(indelParams) == 1, "Only one indel parameter set is supported for indel model"
    indelParams = indelParams[0]
    subRate, rootProb = mixture[0]
    subll = likelihood.subLogLike (seqs, distanceToParent, parentIndex, subRate, rootProb)
    subll_total = float (jnp.sum (subll))

    transMat = jnp.stack ([h20.dummyRootTransitionMatrix()] + [h20.transitionMatrix(t,indelParams,alphabetSize=len(alphabet)) for t in distanceToParent[1:]], axis=0)
    transMat = jnp.log (jnp.maximum (transMat, h20.smallest_float32))
    transll = transCounts * transMat
    transll_total = float (jnp.sum (transll))

    print (json.dumps({'loglike':{'subs':subll_total,'indels':transll_total}, 'cigartree': ct}))

if __name__ == '__main__':
    CLI(main, parser_mode='jsonnet')